crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -1030,8 +1050,31 @@ where
                         self.try_make_sync_target_canonical(num_hash);
                     }
                     InsertPayloadOk::Inserted(BlockStatus::Disconnected { missing_parent }) => {
-                        // continue downloading the missing parent
-                        self.sync.download_full_block(missing_parent.hash);
+                        // compare the missing parent with the canonical tip
+                        let canonical_tip_num = self.blockchain.canonical_tip().number;
+
+                        // if the number of missing blocks is greater than the max, run the
+                        // pipeline
+                        if missing_parent.number >= canonical_tip_num &&
+                            missing_parent.number - canonical_tip_num >
+                                self.pipeline_run_threshold
+                        {
+                            if let Some(target) = self.forkchoice_state_tracker.sync_target() {
```

Comment:
this uses head hash, I think we should consider using finalized or safe instead

we should should also check that this target is not yet canonicalized == missing